In [51]:
# Import required libraries and dependencies
import pandas as pd
import hvplot.pandas
from sklearn.cluster import KMeans
from sklearn.decomposition import PCA
from sklearn.preprocessing import StandardScaler
import numpy as np
import json

import warnings
warnings.filterwarnings("ignore")

In [52]:
# Load the data into a Pandas DataFrame
df_market_data = pd.read_csv(
    "Resources/crypto_market_data.csv",
    index_col="coin_id")

# Display sample data
df_market_data.head(10)

price_change_percentage_24h  price_change_percentage_7d  \
coin_id                                                                    
bitcoin                              1.08388                     7.60278   
ethereum                             0.22392                    10.38134   
tether                              -0.21173                     0.04935   
ripple                              -0.37819                    -0.60926   
bitcoin-cash                         2.90585                    17.09717   
binancecoin                          2.10423                    12.85511   
chainlink                           -0.23935                    20.69459   
cardano                              0.00322                    13.99302   
litecoin                            -0.06341                     6.60221   
bitcoin-cash-sv                      0.92530                     3.29641   

                 price_change_percentage_14d  price_change_percentage_30d  \
coin_id                                                                     
bitcoin                              6.57509                      7.67258   
ethereum                             4.80849                      0.13169   
tether                               0.00640                     -0.04237   
ripple                               2.24984                      0.23455   
bitcoin-cash                        14.75334                     15.74903   
binancecoin                          6.80688                      0.05865   
chainlink                            9.30098                    -11.21747   
cardano                              5.55476                     10.10553   
litecoin                             7.28931                      1.21662   
bitcoin-cash-sv                     -1.86656                      2.88926   

                 price_change_percentage_60d  price_change_percentage_200d  \
coin_id                                                                      
bitcoin                             -3.25185                      83.51840   
ethereum                           -12.88890                     186.77418   
tether                               0.28037                      -0.00542   
ripple                             -17.55245                      39.53888   
bitcoin-cash                       -13.71793                      21.66042   
binancecoin                         36.33486                     155.61937   
chainlink                          -43.69522                     403.22917   
cardano                            -22.84776                     264.51418   
litecoin                           -17.23960                      27.49919   
bitcoin-cash-sv                    -24.87434                       7.42562   

                 price_change_percentage_1y  
coin_id                                      
bitcoin                            37.51761  
ethereum                          101.96023  
tether                              0.01954  
ripple                            -16.60193  
bitcoin-cash                       14.49384  
binancecoin                        69.69195  
chainlink                         325.13186  
cardano                           156.09756  
litecoin                          -12.66408  
bitcoin-cash-sv                    93.73082

In [53]:
# Generate summary statistics
df_market_data.describe()

price_change_percentage_24h  price_change_percentage_7d  \
count                    41.000000                   41.000000   
mean                     -0.269686                    4.497147   
std                       2.694793                    6.375218   
min                     -13.527860                   -6.094560   
25%                      -0.608970                    0.047260   
50%                      -0.063410                    3.296410   
75%                       0.612090                    7.602780   
max                       4.840330                   20.694590   

       price_change_percentage_14d  price_change_percentage_30d  \
count                    41.000000                    41.000000   
mean                      0.185787                     1.545693   
std                       8.376939                    26.344218   
min                     -18.158900                   -34.705480   
25%                      -5.026620                   -10.438470   
50%                       0.109740                    -0.042370   
75%                       5.510740                     4.578130   
max                      24.239190                   140.795700   

       price_change_percentage_60d  price_change_percentage_200d  \
count                    41.000000                     41.000000   
mean                     -0.094119                    236.537432   
std                      47.365803                    435.225304   
min                     -44.822480                     -0.392100   
25%                     -25.907990                     21.660420   
50%                      -7.544550                     83.905200   
75%                       0.657260                    216.177610   
max                     223.064370                   2227.927820   

       price_change_percentage_1y  
count                   41.000000  
mean                   347.667956  
std                   1247.842884  
min                    -17.567530  
25%                      0.406170  
50%                     69.691950  
75%                    168.372510  
max                   7852.089700

In [54]:
# Plot your data to see what's in your DataFrame
df_market_data.hvplot.line(
    width=800,
    height=400,
    rot=90
)

:NdOverlay   [Variable]
   :Curve   [coin_id]   (value)

---

### Prepare the Data

In [55]:
# Use the `StandardScaler()` module from scikit-learn to normalize the data from the CSV file
Scaled_Crypto_Data = StandardScaler().fit_transform(df_market_data)

In [89]:
# Create a DataFrame with the scaled data
Scaled_CD_DF = pd.DataFrame(Scaled_Crypto_Data, columns= df_market_data.columns)

# Copy the crypto names from the original data
Scaled_CD_DF['coin_id'] = df_market_data.index

# Set the coinid column as index
Scaled_CD_DF.set_index('coin_id', inplace=True)

# Display sample data
Scaled_CD_DF.head()

price_change_percentage_24h  price_change_percentage_7d  \
coin_id                                                                 
bitcoin                          0.508529                    0.493193   
ethereum                         0.185446                    0.934445   
tether                           0.021774                   -0.706337   
ripple                          -0.040764                   -0.810928   
bitcoin-cash                     1.193036                    2.000959   

              price_change_percentage_14d  price_change_percentage_30d  \
coin_id                                                                  
bitcoin                          0.772200                     0.235460   
ethereum                         0.558692                    -0.054341   
tether                          -0.021680                    -0.061030   
ripple                           0.249458                    -0.050388   
bitcoin-cash                     1.760610                     0.545842   

              price_change_percentage_60d  price_change_percentage_200d  \
coin_id                                                                   
bitcoin                         -0.067495                     -0.355953   
ethereum                        -0.273483                     -0.115759   
tether                           0.008005                     -0.550247   
ripple                          -0.373164                     -0.458259   
bitcoin-cash                    -0.291203                     -0.499848   

              price_change_percentage_1y  
coin_id                                   
bitcoin                        -0.251637  
ethereum                       -0.199352  
tether                         -0.282061  
ripple                         -0.295546  
bitcoin-cash                   -0.270317

---

### Find the Best Value for k Using the Original Data.

In [57]:
# Create a list with the number of k-values from 1 to 11
k_values = list(range(1, 12))

In [58]:
# Create an empty list to store the inertia values
inertia_values = []

# Create a for loop to compute the inertia with each possible value of k
for k in k_values:
    
# Inside the loop:
# 1. Create a KMeans model using the loop counter for the n_clusters
    kmeans_model = KMeans(n_clusters=k, random_state=1)
# 2. Fit the model to the data using `df_market_data_scaled`
    kmeans_model.fit(Scaled_CD_DF)
# 3. Append the model.inertia_ to the inertia list
    inertia_values.append(kmeans_model.inertia_)

In [59]:
# Create a dictionary with the data to plot the Elbow curve
elbow_data = {'k': k_values, 'inertia': inertia_values}

# Create a DataFrame with the data to plot the Elbow curve
Elbow_df = pd.DataFrame(elbow_data)
Elbow_df 

k     inertia
0    1  287.000000
1    2  195.820218
2    3  123.190482
3    4   79.022435
4    5   65.405923
5    6   52.933559
6    7   47.983124
7    8   37.288187
8    9   33.061685
9   10   28.779752
10  11   25.248499

In [60]:
# Plot a line chart with all the inertia values computed with 
# the different values of k to visually identify the optimal value for k.
Elbow_df.hvplot.line(
    x="index",
    y="inertia",
    title="Elbow Curve",
    xlabel="Number of Clusters (k)",
    ylabel="Inertia",
    xticks=k
)

:Curve   [index]   (inertia)

#### Answer the following question: 

**Question:** What is the best value for `k`?

**Answer:** About 4 clusters

---

### Cluster Cryptocurrencies with K-means Using the Original Data

In [61]:
# Initialize the K-Means model using the best value for k
kmeans_model = KMeans(n_clusters=4)

In [62]:
# Fit the K-Means model using the scaled data
kmeans_model.fit(Scaled_CD_DF)

KMeans(n_clusters=4)

In [63]:
# Predict the clusters to group the cryptocurrencies using the scaled data
crypto_clusters = kmeans_model.predict(Scaled_CD_DF)

# Print the resulting array of cluster values.
print(crypto_clusters)

[0 0 2 2 0 0 0 0 0 2 2 2 2 0 2 0 2 2 0 2 2 0 2 2 2 2 2 2 0 2 2 2 1 0 2 2 3
 2 2 2 2]


In [64]:
# Create a copy of the DataFrame
clusters_df = Scaled_CD_DF.copy()

In [67]:
# Add a new column to the DataFrame with the predicted clusters
clusters_df["cluster"] = crypto_clusters

clusters_df.head()

price_change_percentage_24h  price_change_percentage_7d  \
coin_id                                                                 
bitcoin                          0.508529                    0.493193   
ethereum                         0.185446                    0.934445   
tether                           0.021774                   -0.706337   
ripple                          -0.040764                   -0.810928   
bitcoin-cash                     1.193036                    2.000959   

              price_change_percentage_14d  price_change_percentage_30d  \
coin_id                                                                  
bitcoin                          0.772200                     0.235460   
ethereum                         0.558692                    -0.054341   
tether                          -0.021680                    -0.061030   
ripple                           0.249458                    -0.050388   
bitcoin-cash                     1.760610                     0.545842   

              price_change_percentage_60d  price_change_percentage_200d  \
coin_id                                                                   
bitcoin                         -0.067495                     -0.355953   
ethereum                        -0.273483                     -0.115759   
tether                           0.008005                     -0.550247   
ripple                          -0.373164                     -0.458259   
bitcoin-cash                    -0.291203                     -0.499848   

              price_change_percentage_1y  cluster  
coin_id                                            
bitcoin                        -0.251637        0  
ethereum                       -0.199352        0  
tether                         -0.282061        2  
ripple                         -0.295546        2  
bitcoin-cash                   -0.270317        0

In [68]:
# Create a scatter plot using hvPlot by setting 
# `x="price_change_percentage_24h"` and `y="price_change_percentage_7d"`. 
# Color the graph points with the labels found using K-Means and 
# add the crypto name in the `hover_cols` parameter to identify 
# the cryptocurrency represented by each data point.

clusters_plot = clusters_df.hvplot.scatter(
    x="price_change_percentage_24h", 
    y="price_change_percentage_7d",
    by= "cluster",
    hover_cols=['coin_id'],
    title="Original Data Plot",
    width=500, height=300
    )
clusters_plot

:NdOverlay   [cluster]
   :Scatter   [price_change_percentage_24h]   (price_change_percentage_7d,coin_id)

---

### Optimize Clusters with Principal Component Analysis.

In [69]:
# Create a PCA model instance and set `n_components=3`.
pca = PCA(n_components=3)

In [74]:
# Use the PCA model with `fit_transform` to reduce to 
# three principal components.
crypto_pca = pca.fit_transform(Scaled_CD_DF)

# View the first five rows of the DataFrame. 
crypto_pca_df = pd.DataFrame(
    data=crypto_pca, 
    columns=["principal component 1", "principal component 2", "principal component 3"],
    index=Scaled_CD_DF.index
)

crypto_pca_df.head()

principal component 1  principal component 2  \
coin_id                                                      
bitcoin                   -0.600667               0.842760   
ethereum                  -0.458261               0.458466   
tether                    -0.433070              -0.168126   
ripple                    -0.471835              -0.222660   
bitcoin-cash              -1.157800               2.041209   

              principal component 3  
coin_id                              
bitcoin                    0.461595  
ethereum                   0.952877  
tether                    -0.641752  
ripple                    -0.479053  
bitcoin-cash               1.859715

In [76]:
# Retrieve the explained variance to determine how much information 
# can be attributed to each principal component.
explained_variance = pca.explained_variance_ratio_
Scope = 0

for i, variance in enumerate(explained_variance):
    Scope = Scope + variance
    print(f"Explained Variance PC{i+1}: {variance:.4f}")

print(f"{Scope * 100:.2f}%")

Explained Variance PC1: 0.3720
Explained Variance PC2: 0.3470
Explained Variance PC3: 0.1760
89.50%


#### Answer the following question: 

**Question:** What is the total explained variance of the three principal components?

**Answer:** The total explained variance is 89.5% of the data, explained by the three component PCA

In [77]:
# Create a new DataFrame with the PCA data.
pca_data = pd.DataFrame(data=crypto_pca, columns=['PC 1', 'PC 2', 'PC 3'])

# Copy the crypto names from the original data
pca_data["coin_id"] = pca_data.index

# Set the coinid column as index
pca_data.set_index("coin_id", inplace=True)

# Display sample data
pca_data.head()

PC 1      PC 2      PC 3
coin_id                              
0       -0.600667  0.842760  0.461595
1       -0.458261  0.458466  0.952877
2       -0.433070 -0.168126 -0.641752
3       -0.471835 -0.222660 -0.479053
4       -1.157800  2.041209  1.859715

---

### Find the Best Value for k Using the PCA Data

In [78]:
# Create a list with the number of k-values from 1 to 11
k_values

[1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11]

In [79]:
# Create an empy list to store the inertia values
inertia = []

# Create a for loop to compute the inertia with each possible value of k
for k in k_values:

    # 1. Create a KMeans model using the loop counter for the n_clusters
    kmeans_model = KMeans(n_clusters=k, random_state=0)

    # 2. Fit the model to the data using `df_market_data_pca`
    kmeans_model.fit(pca_data)
    # 3. Append the model.inertia_ to the inertia list
    inertia.append(kmeans_model.inertia_)

In [80]:
# Create a dictionary with the data to plot the Elbow curve
elbow_data_2 = {"k": k_values, "inertia": inertia}

# Create a DataFrame with the data to plot the Elbow curve
elbow_df_2 = pd.DataFrame(elbow_data_2)

elbow_df_2

k     inertia
0    1  256.874086
1    2  165.901994
2    3   93.774626
3    4   49.665497
4    5   37.878747
5    6   27.618972
6    7   21.182776
7    8   17.389823
8    9   13.593650
9   10   10.559358
10  11    8.007418

In [81]:
# Plot a line chart with all the inertia values computed with 
# the different values of k to visually identify the optimal value for k.
elbow_curve_2 = elbow_df_2.hvplot.line(x='k', y='inertia', xlabel='Number of clusters (k)', 
                                       ylabel='Inertia', title='Elbow Curve - PCA Data', 
                                       color='red', width=500, height=400)

result_2 = elbow_curve_2 * elbow_df_2.hvplot.scatter(x='k', y='inertia', marker='x', color='red', size=75)
result_2

:Overlay
   .Curve.I   :Curve   [k]   (inertia)
   .Scatter.I :Scatter   [k]   (inertia)

#### Answer the following questions: 

* **Question:** What is the best value for `k` when using the PCA data?

  * **Answer:** The best value for 'k' when using PCA data is 4 clusters


* **Question:** Does it differ from the best k value found using the original data?

  * **Answer:** No, the elbow curves are very similar and the optimal value for 'k' is the same 

### Cluster Cryptocurrencies with K-means Using the PCA Data

In [82]:
# Initialize the K-Means model using the best value for k
kmeans_model_2 = KMeans(n_clusters=4)

In [83]:
# Fit the K-Means model using the PCA data
kmeans_model_2.fit(pca_data)

KMeans(n_clusters=4)

In [84]:
# Predict the clusters to group the cryptocurrencies using the PCA data
clusters_2 = kmeans_model_2.predict(pca_data)

# Print the resulting array of cluster values.
print(clusters_2)

[1 1 3 3 1 1 1 1 1 3 3 3 3 1 3 1 3 3 1 3 3 1 3 3 3 3 3 3 1 3 3 3 2 1 3 3 0
 3 3 3 3]


In [85]:
# Create a copy of the DataFrame with the PCA data
clusters_df_2 = pca_data.copy()

# Add a new column to the DataFrame with the predicted clusters
clusters_df_2["cluster"] = clusters_2

# Display sample data
clusters_df_2.head()

PC 1      PC 2      PC 3  cluster
coin_id                                       
0       -0.600667  0.842760  0.461595        1
1       -0.458261  0.458466  0.952877        1
2       -0.433070 -0.168126 -0.641752        3
3       -0.471835 -0.222660 -0.479053        3
4       -1.157800  2.041209  1.859715        1

In [87]:
# Create a scatter plot using hvPlot by setting 
# `x="price_change_percentage_24h"` and `y="price_change_percentage_7d"`. 
# Color the graph points with the labels found using K-Means and 
# add the crypto name in the `hover_cols` parameter to identify 
# the cryptocurrency represented by each data point.
clusters_plot_2 = clusters_df_2.hvplot.scatter(
    x="PC 1", 
    y="PC 2",
    by= "cluster",
    hover_cols=['coin_id'],
    xlabel='price_change_percentage_24h',
    ylabel='price_change_percentage_7d',
    title="PCA Data Plot",
    width=500, height=300
    )

clusters_plot_2

:NdOverlay   [cluster]
   :Scatter   [PC 1]   (PC 2,coin_id)

### Visualize and Compare the Results

In this section, you will visually analyze the cluster analysis results by contrasting the outcome with and without using the optimization techniques.

In [90]:
# Composite plot to contrast the Elbow curves


In [31]:
# Composite plot to contrast the clusters
# YOUR CODE HERE!

#### Answer the following question: 

  * **Question:** After visually analyzing the cluster analysis results, what is the impact of using fewer features to cluster the data using K-Means?

  * **Answer:** After visually analyzing the cluster analysis results, using fewer clusters with the PCA data, the results are accurately clustered. With outliers compared to using more clusters, two clusters were really close which can not represent the data. 